### ISSUE TITLE
Question about (padding) masking in cross-attention with transformers

### ISSUE THREAD

**Issue by m-decoster** [joeynmt/joeynmt#151]
In the decoder of the transformer model, we apply cross-attention between the "memory" (encoder outputs) and "targets" (decoder inputs).

For this, in the `TransformerDecoderLayer`, we use `src_mask` as mask: https://github.com/joeynmt/joeynmt/blob/master/joeynmt/transformer_layers.py#L269

This mask is of the shape `(batch_size, 1, source_length)` and in `MultiHeadedAttention` it is reshaped to `(batch_size, 1, 1, source_length)`: https://github.com/joeynmt/joeynmt/blob/master/joeynmt/transformer_layers.py#L269 which allows for broadcasting for multi-head attention.

The source sentence padding is masked, but the target sentence is not. I had a batch with several sentences of different lengths, so some shorter (target) sentences had padding tokens. I noticed when I visualized the attention scores, that the model was attending to these padding tokens.

Is this intended behavior? Shouldn't the mask be of shape `(batch_size, target_length, source_length)` to avoid attending to padding in the target sentence? Notice for example how the PyTorch implementation of MultiHeadAttention has both a mask and a padding mask, which I think achieves the intended behavior: https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L4659

**Comment by may-** [comment]
Hi,

do you visualize the attention in inference time? In inference time, the model decodes the sequence one by one, by applying the subsequent_mask in order not to attend future tokens.  https://github.com/joeynmt/joeynmt/blob/42ad5889c622e2583b667cc58bc7c37f3c99111c/joeynmt/decoders.py#L535
what exactly you mean by `target_length`? We don't know the target length in inference time. maybe the steps so far unrolled?

I thought the current implementation is correct, but not sure, honestly.

**Comment by m-decoster** [comment]
After some more reasoning, I think this isn't actually a problem because the loss function ignores any padding tokens in the targets.

**Comment by may-** [comment]
I think both `src_mask` and `trg_mask` has `shape = (batch_size, 1, src_len/trg_len)` and it should be, for src_src attention in encoder, and for trg_trg attention in decoder. So the problem is, we probably need so to called a `memory_mask` (just a random name for now) which has `shape = (batch_size, trg_len, src_len)` for src_trg attention. Have I understand your problem correctly?
